File: Algorithm/147. Insertion Sort List.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-07-30 11:00:18
# @Last modified by:   WuLC
# @Last Modified time: 2016-07-30 11:14:28
# @Email: liangchaowu5@gmail.com


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# method 1,directly implement insertion sort on the linked list,O(n^2),TLE
class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None: return None
        p1 = head.next
        dummy = ListNode(0)
        dummy.next, end, end.next = head, head, None
        while p1:
            p2 = p1.next
            tmp1, tmp2 = dummy, dummy.next
            while tmp2:
                if p1.val < tmp2.val:
                    tmp1.next = p1
                    p1.next = tmp2
                    break
                else:
                    tmp1 = tmp1.next
                    tmp2= tmp2.next
            if tmp2 == None:
                end.next = p1
                end = end.next
                end.next = None
            p1 = p2
        return dummy.next
        
        
# method 2, copy the values of the linked list to an array and sort the array, O(nlogn), AC
class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        tmp,values = head, []
        while tmp:
            values.append(tmp.val)
            tmp = tmp.next
        tmp = head
        values.sort()
        for i in xrange(len(values)):
            tmp.val = values[i]
            tmp = tmp.next
        return head
            

File: Algorithm/Java/147. Insertion Sort List.java
/*
 * Created on Fri Nov 10 2017 14:28:16
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// Insertion sort
// O(n^2) time, O(1) space


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution 
{
    public ListNode insertionSortList(ListNode head) 
    {
        if (head == null || head.next == null) return head;
        ListNode curr = head.next, p;
        while(curr != null)
        {
            p = head;
            int tmp = curr.val;
            while(p != curr)
            {
                if (tmp <= p.val)
                {
                    int t = p.val;
                    p.val = tmp;
                    tmp = t;
                }
                p = p.next;
            }
            p.val = tmp;
            curr = curr.next;
        }
        return head;
    }
}
